### PR TITLE
[PR] socket io integration

### PR DESCRIPTION
**Pull request by tosinamuda**

### CODE REVIEW

File: pages/user.js [hunk 1]
```diff
@@ -1,10 +1,22 @@
+/* eslint-disable no-unused-vars */
```
Comment (enyeInc) [inline pages/user.js]: remove this disable

File: pages/user.js [hunk 2]
```diff
@@ -1,10 +1,22 @@
...
 import React, { Component } from 'react';
 import { Row, Col } from 'antd';
 import fetch from 'isomorphic-unfetch';
+import io from 'socket.io-client';
 import Layout from '../components/layout';
 import './admin/profile/index.scss';
+import SOCKET_IO_EVENTS from '../server/lib/config/constants';
```
Comment (enyeInc) [inline pages/user.js]: refer to the coding convention for how to order these imports

File: pages/user.js [hunk 3]
```diff
@@ -1,10 +1,22 @@
...
 
 class User extends Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            hello: '',
```
Comment (enyeInc) [inline pages/user.js]: what is this `hello` key?

File: pages/user.js [hunk 4]
```diff
@@ -34,9 +46,67 @@ class User extends Component {
...
+            if (connected) this.socket.emit(message, { text });
+        };
+        const leaveChat = () => {
+            // call this when user want to end current chat
+            if (connected) {
+                this.socket.emit(leaveRoom);
+                this.socket.leave(room);
+                room = '';
+            }
+        };
```
Comment (enyeInc) [inline pages/user.js]: make `sendMessage` and `leaveChat` methods on the `User` class

File: pages/user.js [hunk 5]
```diff
@@ -49,6 +119,7 @@ class User extends Component {
                                 src={isLoggedIn ? picture : ''}
                             />
                         </div>
+                        <h1>{hello}</h1>
```
Comment (enyeInc) [inline pages/user.js]: Again what is the use of this `hello` variable?

File: server/lib/config/passport.js [hunk 6]
```diff
@@ -6,16 +6,13 @@ const User = require('../../models/User');
...
     AUTH0_CLIENT_ID,
     AUTH0_CLIENT_SECRET,
     AUTH0_DOMAIN,
     AUTH0_CALLBACK_URL,
 } = process.env;
 const callbackURL =
-	NODE_ENV === 'production'
-	    ? AUTH0_CALLBACK_URL
-	    : 'http://localhost:5000/api/users/callback';
+	AUTH0_CALLBACK_URL || 'http://localhost:5000/api/users/callback';
```
Comment (enyeInc) [inline server/lib/config/passport.js]: `http://localhost:5000/api/users/callback` should be saved to a variable

File: server/lib/config/socketio.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    challengeEnd,
+    challengeStart,
+    connection,
+    disconnet,
+    getUser,
+    leaveRoom,
+    login,
+    message,
+    users,
+} = SOCKET_IO_EVENTS;
```
Comment (enyeInc) [inline server/lib/config/socketio.js]: `const {...} = CONFIG_CONSTANTS. SOCKET_IO_EVENTS` is a bit cleaner

File: server/lib/config/socketio.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            peer.emit(challengeStart, { name: names[socket.id], room });
+            socket.emit(challengeStart, { name: names[peer.id], room });
+        } else {
+            // queue is empty, add our lone socket
+            queue.push(socket);
+        }
+    };
+
+    io.on(connection, socket => {
+        console.log(`User ${socket.id} connected`);
```
Comment (enyeInc) [inline server/lib/config/socketio.js]: remove console.log

File: server/lib/config/socketio.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            socket.broadcast.to(room).emit(message, data);
+        });
+        socket.on(leaveRoom, () => {
+            const room = rooms[socket.id];
+            socket.broadcast.to(room).emit(challengeEnd);
+            let peerID = room.split('#');
+            peerID = peerID[0] === socket.id ? peerID[1] : peerID[0];
+            // add both current and peer to the queue
+            findPeerForLoneSocket(allUsers[peerID]);
+            findPeerForLoneSocket(socket);
```
Comment (enyeInc) [inline server/lib/config/socketio.js]: a bit cleaner
```
[allUsers[peerID], socket].forEach(item => findPeerForLoneSocket(item))
```

File: server/server.js [hunk 10]
```diff
@@ -1,4 +1,8 @@
 const express = require('express');
+
+const server = express();
+const httpServer = require('http').Server(server);
+const io = require('socket.io')(httpServer);
 const next = require('next');
 const bodyParser = require('body-parser');
 const mongoose = require('mongoose');
```
Comment (enyeInc) [inline server/server.js:9]: consult the coding conventions on how to order these imports

**Comment by enyeInc** [comment]
@tosinamuda From what I understood from your code, you are matching a user with the next person in the queue. Okay that works, but think about how your code will change if you want to challenge a specific person. You do not have to add that to this PR, but your next once should address it. Here is a hint you will need to use [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) or [Maps](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) to store your data

**Comment by enyeInc** [comment]
@tosinamuda Also look at this stackoverflow page about how to all connected sockets in the newest version off socket.io (https://stackoverflow.com/questions/6563885/socket-io-how-do-i-get-a-list-of-connected-sockets-clients)

File: server/server.js [hunk 11]
```diff
@@ -88,7 +91,11 @@ app.prepare()
         server.use('/', routes);
 
         server.get('*', (req, res) => handle(req, res));
-        server.listen(_PORT, err => {
+
+        // eslint-disable-next-line global-require
+        require('./lib/config/socketio')(io);
```
Comment (teenoh) [inline server/server.js]: why isn't the require function at the top of the file

File: pages/user.js (hunk 5 above)
Comment (tosinamuda) [reply pages/user.js]: I was using it initially for debugging

File: pages/user.js (hunk 2 above)
Comment (tosinamuda) [reply pages/user.js]: Not sure there is something in the coding convention regarding the import

File: server/server.js (hunk 10 above)
Comment (tosinamuda) [reply server/server.js:9]: @enyeInc I separated the require from initialization, is that ok? I couldnt find anything in the doc on this. And the doc is more inclined to the frontend as opposed to backend
```
const express = require('express');
const http = require('http');
const socketio = require('socket.io');
const next = require('next');
const bodyParser = require('body-parser');
const mongoose = require('mongoose');
const morgan = require('morgan');
const passport = require('passport');
const cookieParser = require('cookie-parser');
const session = require('express-session');
const dotenv = require('dotenv');
const MongoStore = require('connect-mongo')(session);
const path = require('path');
const userInViews = require('./lib/middlewares/userInViews');
const routes = require('./routes');
```

File: pages/user.js (hunk 2 above)
Comment (enyeInc) [reply pages/user.js]: You are right, I forgot to add it. Here is it though
```
// external dependencies in alpha-numeric order

// internal dependencies in alpha-numeric order

// constant declarations
```

File: server/server.js (hunk 10 above)
Comment (enyeInc) [reply server/server.js:9]: i added a comment on how to address this

File: pages/user.js (hunk 3 above)
Comment (Way2nnadi) [reply pages/user.js]: You have to answer the question before you resolve the comment

File: server/server.js (hunk 11 above)
Comment (tosinamuda) [reply server/server.js]: It is like a method to abstract the socket config and should be called at this point but I have  found a  better way to handle this

File: pages/user.js (hunk 3 above)
Comment (tosinamuda) [reply pages/user.js]: This is an oversight, I mentioned in another comment which also referred to " hello", I used it to test sending data through socket and shouldn't have pushed it

File: server/lib/config/passport/constants.js [hunk 12]
```diff
@@ -0,0 +1,5 @@
+const PASSPORT_AUTH0_CONFIG = {
+    fallbackCallback: 'http://localhost:5000/api/users/callback',
```
Comment (Way2nnadi) [inline server/lib/config/passport/constants.js]: `FALL_BACK_URL = '/api/users/callback';`

File: server/lib/config/passport/index.js [hunk 13]
```diff
@@ -1,21 +1,19 @@
...
     AUTH0_CLIENT_ID,
     AUTH0_CLIENT_SECRET,
     AUTH0_DOMAIN,
     AUTH0_CALLBACK_URL,
 } = process.env;
 const callbackURL =
-	NODE_ENV === 'production'
-	    ? AUTH0_CALLBACK_URL
-	    : 'http://localhost:5000/api/users/callback';
+	AUTH0_CALLBACK_URL || PASSPORT_AUTH0_CONFIG.fallbackCallback;
```
Comment (Way2nnadi) [inline server/lib/config/passport/index.js]: `AUTH0_CALLBACK_URL || FALL_BACK_URL`

File: pages/user.js (hunk 1 above)
Comment (Way2nnadi) [reply pages/user.js]: This still has to be removed, you can not have unused variables in your files

File: pages/user.js [hunk 14]
```diff
@@ -150,7 +150,10 @@ class User extends Component {
             <Button
                 type="primary"
                 size="small"
-                onClick={() => this.acceptChallengeRequest()}
+                onClick={() => {
+                    this.acceptChallengeRequest();
+                    notification.close(key);
+                }}
```
Comment (Way2nnadi) [inline pages/user.js]: create a method on the user class called `onButtonClick` and pass it as a property to `Button`
```
...
onClick={this.onButtonClick}
...
```
You might need to bind the function in the constructor.

File: pages/user.js [hunk 15]
```diff
@@ -34,9 +71,127 @@ class User extends Component {
...
+        );
+        notification.open({
+            btn,
+            description: `${user.name} has challenged you`,
+            duration: 0,
+            key,
+            message: 'Challenge Notification',
+            onClose: () => {
+                this.rejectChallengeRequest();
+            },
```
Comment (Way2nnadi) [inline pages/user.js]: `onClose: {this.rejectChallengeRequest}`

File: pages/user.js [hunk 16]
```diff
@@ -34,9 +71,127 @@ class User extends Component {
...
+        });
+    }
+
+    challengeUser(selectedUser) {
```
Comment (Way2nnadi) [inline pages/user.js:169]: destructure `selectedUser`

**Review by Way2nnadi — APPROVED** [review]
Minor changes but looks good to me

File: pages/user.js [hunk 17]
```diff
@@ -34,9 +71,125 @@ class User extends Component {
...
+            this.setState({ room: '' });
+        }
+    }
+
+    handleSocket() {
+        const { connect } = CLIENT_SYSTEM_EVENTS;
+        const { userData: user, isLoggedIn } = this.props;
+        const { room } = this.state;
+        const { _id, name, picture } = user;
+        this.socket = io();
```
Comment (Way2nnadi) [inline pages/user.js]: put `this.socket = io();` in the constructor

File: pages/user.js [hunk 18]
```diff
@@ -34,9 +71,125 @@ class User extends Component {
...
+
+            if (isLoggedIn) {
+                this.socket.emit(login, { _id, name, picture });
+            }
+        });
+
+        this.socket.on(users, data => {
+            this.setState({ onlineUsers: data.users });
+            const { onlineUsers } = this.state;
+            console.log(`users called ${onlineUsers.length}`);
```
Comment (Way2nnadi) [inline pages/user.js]: remove all `console.log`

File: pages/user.js [hunk 19]
```diff
@@ -34,9 +71,125 @@ class User extends Component {
...
+
+        this.socket.on(challengeRequest, data => {
+            const { room: roomId, user: sender } = data;
+            this.setState({ room: roomId });
+            this.openNotification(sender);
+            // TODO show notification to user; challengeRequestNotification(data.user);
+        });
+
+        this.socket.on(challengeEnd, () => {
+            // TODO show notification to user; challengeEndNotification();
```
Comment (Way2nnadi) [inline pages/user.js]: remove all these separate `TODO`'s, add one todo at the top of `handleSocket` saying that you need to figure out how to `show notification to user`

File: pages/user.js [hunk 20]
```diff
@@ -34,9 +71,125 @@ class User extends Component {
...
+            this.socket.leave(room);
+            this.setState({ room: '' });
+        });
+    }
+
+    // eslint-disable-next-line class-methods-use-this
```
Comment (Way2nnadi) [inline pages/user.js]: why is eslint disabled?

File: pages/user.js [hunk 21]
```diff
@@ -49,6 +202,22 @@ class User extends Component {
...
+                                    <Button
+                                        type="Primary"
+                                        onClick={() =>
+                                            this.challengeUser(value)
+                                        }
+                                    >
+										Challenge
+                                    </Button>
+                                </li>
+                            ))}
```
Comment (Way2nnadi) [inline pages/user.js]:
```
{
    onlineUsers.map(...)
}
```

File: pages/user.js (hunk 18 above)
Comment (teenoh) [reply pages/user.js]: This won't be added to our code. `pages/user.js` is just a test file tosin works with. I will need it as a guide when I'm properly implementing the code there.

Comment (teenoh) [reply pages/user.js]: tl;dr I need the console.logs

**Review by Way2nnadi — APPROVED** [review]
Minor changes, but looks good overall

File: pages/user.js (hunk 21 above)
Comment (tosinamuda) [reply pages/user.js]: I didnt quite get this
Do you mean making it a one liner like
` {
     onlineUsers.map(value => displayOnlineUser(value))
}`

Eslint is not going to allow me do this it removes the whitespace

File: pages/user.js (hunk 20 above)
Comment (tosinamuda) [reply pages/user.js]: Eslint expects me to use atleast one this.something call in that method, I was not doing this before, that was why I opted for a disable

**Review by teenoh — APPROVED** [review]

**Pull request merged**